tracks/rust/exercises/bob/mentoring.md

```diff
@@ -0,0 +1,140 @@
+- [Iterator Methods](https://doc.rust-lang.org/std/iter/trait.Iterator.html#provided-methods)
+- [constant](https://doc.rust-lang.org/std/keyword.const.html)
+
+### Reasonable Solutions
+
+A reasonable solution should:
+
+- Use constants
+- Use variables
+- Use `match` statements
+- Check if the input string is empty and return the "Sure" constant
+before doing any additional processing
+- Use simple methods from the standard library
+- Not use the regex crate
+- Not use separate functions
+
+#### Examples
+
+```rust
+const SURE: &str = "Sure.";
+const WHOA: &str = "Whoa, chill out!";
+const CALM: &str = "Calm down, I know what I'm doing!";
+const FINE: &str = "Fine. Be that way!";
+const WHATEVER: &str = "Whatever.";
+
+#[allow(clippy::match_bool)]
+pub fn reply(message: &str) -> &str {
+    let trimmed: &str = message.trim();
+
+    match trimmed.is_empty() {
+        true => FINE,
+        false => {
+            let is_question: bool = trimmed.ends_with("?");
+
+            let contains_alphabetic_characters: bool = trimmed.chars().any(|character| character.is_alphabetic());
+            let is_uppercase: bool = trimmed == trimmed.to_uppercase();
+            let is_yelled: bool = contains_alphabetic_characters && is_uppercase;
+
+            match (is_question, is_yelled) {
+                (true, false) => SURE,
+                (false, true) => WHOA,
+                (true, true) => CALM,
+                (false, false) => WHATEVER,
+            }
+        }
+    }
+}
+```
+
+### Common Suggestions
+
+If they don't use a constant:
+```
+We can improve this by setting our output strings to be [constants](https://doc.rust-lang.org/std/keyword.const.html)
+```
+
+If they don't create a trimmed variable or use something other than `.trim()`:
+```
+Have a look at the `.trim()` method.
+```
+
+If they check if the string is empty in any other way than `.is_empty()`:
+```
+Well done you for seeing that we need to check if the string is empty with `== ""`. There's actually a method, `.is_empty()`, that we can use instead, which is much cleaner.
+```
+
+If they don't use `match`:
+```
+Check out using `match` instead of `if`. For some things, like destructuring, it's required. It's an idiom specific to Rust, so I always default to it unless the logic is comparing many different things and a match in that case makes it messy.
+```
+
+If they use an `if` on their `.is_empty()` check:
+```
+See about doing a  `match`, instead of `if`, on `your_trimmed_variable.is_empty()`. And, if it's `true`, instantly output our "Fine. Be that way!" constant. Then, everything else will be in the `false` arm.
+Like this:
+    ```rust
+    match your_trimmed_variable.is_empty() {
+        true => FINE,
+        false => {
+    ```
+Be aware that if you use Clippy, it will tell you that it is improper, it's purely a stylistic choice :)
+We can tell Clippy to allow it with `#[allow(clippy::match_bool)]`.
+```
```

Comment:
This line needs to be removed, doesn't it?

Reply:
The issue is the nested code block inside the comment.

@TheDarkula  you can use 4 space identation for line 82-92 (remove the backticks) and then the backticked code block will remain.

Reply:
Personally I prefer to use longer fences when nesting codeblocks instead of increasing indentation:

`````md
````md
This has another codeblock:

```rs
pub mod foo;
```
````
`````

Which will render like this:

````md
This has another codeblock:

```rs
pub mod foo;
```
````

Reply:
Ah yes @NobbZ . I like that!

Reply:
Good catch, thanks so much! Got it changed here and for lines 116-121 :)